File: sdks/php/src/Crypto/JWKThumbprint.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Crypto;

/**
 * JWK Thumbprint computation (RFC 7638) for Ed25519 keys.
 */
final class JWKThumbprint
{
    /**
     * Compute the JWK thumbprint for an Ed25519 public key.
     *
     * @param string $publicKey 32-byte raw public key
     * @return string Base64url-encoded SHA-256 thumbprint
     */
    public static function compute(string $publicKey): string
    {
        $x = Base64Url::encode($publicKey);

        // Canonical JSON with sorted keys, no whitespace
        $canonical = json_encode([
            'crv' => 'Ed25519',
            'kty' => 'OKP',
            'x' => $x,
        ], JSON_UNESCAPED_SLASHES);

        $digest = hash('sha256', $canonical, true);

        return Base64Url::encode($digest);
    }
}

File: sdks/java/src/main/java/dev/atap/sdk/crypto/JWKThumbprint.java
package dev.atap.sdk.crypto;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * JWK Thumbprint computation (RFC 7638) for Ed25519 public keys.
 */
public final class JWKThumbprint {

    private JWKThumbprint() {
    }

    /**
     * Compute the JWK thumbprint for an Ed25519 public key.
     *
     * @param publicKeyBytes the 32-byte Ed25519 public key
     * @return base64url-encoded SHA-256 thumbprint
     */
    public static String compute(byte[] publicKeyBytes) {
        String x = Base64Url.encode(publicKeyBytes);
        // Canonical JSON: sorted keys, no whitespace
        String canonical = "{\"crv\":\"Ed25519\",\"kty\":\"OKP\",\"x\":\"" + x + "\"}";
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(canonical.getBytes(StandardCharsets.UTF_8));
            return Base64Url.encode(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }
}
